From f4d4791ab01ec567fca319a3b0474edfb79f00cf Mon Sep 17 00:00:00 2001
From: Xie XiuQi <xiexiuqi@huawei.com>
Date: Mon, 1 Apr 2019 11:59:09 +0800
Subject: [PATCH 2067/2592] [Backport] sched/cputime: use sched idle time
 accounting
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

hulk inclusion
category: bugfix
bugzilla: 13257

Reference: https://gitee.com/openeuler/kernel/commit/1af75aa95dc59c902a0ce961fa499b44a2fdf97c

-------------------------------------------------

Interduce "use-sched-idle-time" option to enable/disable
sched idle time accounting instead of tick based time accounting.

nohz mode: use get_idle_time;
nohz=off mode: use idle time accounting in sched_info_switch function.

And you can also use "use-sched-idle-time=force" force to use
idle time accountint in sched_info_switch function.

guarantee:
cpu_clock(cpu) - idle == utime + stime

We use this time in /proc/stat.

We use get_idle_time instead of collecting idle time when
context switch. And get_idle_time is available when nohz is
enabled. So do not collect idle time when nohz is enabled.

Issue 1:
1) boot euleros
2) bind a kernel thread to a cpu (eg. cpu1), run a long time (more than 10mins)
3) stop this kernel thread
4) run a spin loop process in userspace
5) then, the utime is 0, but the stime is ~100%

for example:
stime = 10000; utime = 1; rtime = 100001;
Run 10s in userspace now, then

rtime = clock - idle , (10001 + 10 = 10011)
stime = 10 * 10000/(1 + 10 + 10000) ≈ 10
utime = 10 * (1 + 10)/(1 + 10 + 10000) ≈ 0

That's the problem.

We just adjust the delta time to reduce the accumulated:
error of stime/utime:
delta = rtime - (utime + stime)
stime += delta * delta_of_stat_stime / (delta_of_stat_stime + delta_of_stat_utime)
utime = rtime - stime

Issue 2:
When cpu is offline we don't need to adjust the user and system time.
Only cpu is online we could asume:

utime% + stime% + idle% = 100%.

Issue 3:
When nohz=off, we must add 'now - last_entry_idle_time' to sum_idle_time.

v2:
 - fix compile error when use allmodconfig.

conflicts:
	kernel/sched/core.c
	kernel/sched/cputime.c
	kernel/sched/sched.h
	kernel/sched/stats.h

Tested-by: Xie Zhipeng <xiezhipeng1@huawei.com>
Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Zheng Zucheng <zhengzucheng@huawei.com>
---
 fs/proc/stat.c                | 25 ++++++++++---
 include/linux/sched/cputime.h |  3 ++
 include/linux/sched/nohz.h    |  2 ++
 kernel/sched/core.c           | 14 ++++++++
 kernel/sched/cputime.c        | 67 +++++++++++++++++++++++++++++++++++
 kernel/sched/idle.c           | 10 ++++++
 kernel/sched/sched.h          | 15 ++++++++
 kernel/time/tick-internal.h   |  3 +-
 8 files changed, 132 insertions(+), 7 deletions(-)

diff --git a/fs/proc/stat.c b/fs/proc/stat.c
index 3bed48d8228b..1b65fadff729 100644
--- a/fs/proc/stat.c
+++ b/fs/proc/stat.c
@@ -52,7 +52,9 @@ u64 get_idle_time(struct kernel_cpustat *kcs, int cpu)
 	if (cpu_online(cpu))
 		idle_usecs = get_cpu_idle_time_us(cpu, NULL);
 
-	if (idle_usecs == -1ULL)
+	if (idle_usecs == -1ULL && static_branch_unlikely(&use_sched_idle_time))
+		return sched_get_idle_time(cpu);
+	else if (idle_usecs == -1ULL)
 		/* !NO_HZ or cpu offline so we can rely on cpustat.idle */
 		idle = kcs->cpustat[CPUTIME_IDLE];
 	else
@@ -125,9 +127,18 @@ static int show_stat(struct seq_file *p, void *v)
 
 		kcpustat_cpu_fetch(&kcpustat, i);
 
-		user		+= cpustat[CPUTIME_USER];
+		if (static_branch_unlikely(&use_sched_idle_time) && cpu_online(i)) {
+			u64 u = 0, s = 0;
+
+			sched_idle_time_adjust(i, &u, &s);
+
+			user += u;
+			system += s;
+		} else {
+			user += cpustat[CPUTIME_USER];
+			system += cpustat[CPUTIME_SYSTEM];
+		}
 		nice		+= cpustat[CPUTIME_NICE];
-		system		+= cpustat[CPUTIME_SYSTEM];
 		idle		+= get_idle_time(&kcpustat, i);
 		iowait		+= get_iowait_time(&kcpustat, i);
 		irq		+= cpustat[CPUTIME_IRQ];
@@ -166,9 +177,13 @@ static int show_stat(struct seq_file *p, void *v)
 		kcpustat_cpu_fetch(&kcpustat, i);
 
 		/* Copy values here to work around gcc-2.95.3, gcc-2.96 */
-		user		= cpustat[CPUTIME_USER];
+		if (static_branch_unlikely(&use_sched_idle_time)) {
+			sched_idle_time_adjust(i, &user, &system);
+		} else {
+			user = cpustat[CPUTIME_USER];
+			system = cpustat[CPUTIME_SYSTEM];
+		}
 		nice		= cpustat[CPUTIME_NICE];
-		system		= cpustat[CPUTIME_SYSTEM];
 		idle		= get_idle_time(&kcpustat, i);
 		iowait		= get_iowait_time(&kcpustat, i);
 		irq		= cpustat[CPUTIME_IRQ];
diff --git a/include/linux/sched/cputime.h b/include/linux/sched/cputime.h
index 6c9f19a33865..e05476ad88ae 100644
--- a/include/linux/sched/cputime.h
+++ b/include/linux/sched/cputime.h
@@ -187,5 +187,8 @@ static inline void prev_cputime_init(struct prev_cputime *prev)
 
 extern unsigned long long
 task_sched_runtime(struct task_struct *task);
+DECLARE_STATIC_KEY_FALSE(use_sched_idle_time);
+extern int sched_idle_time_adjust(int cpu, u64 *utime, u64 *stime);
+extern unsigned long long sched_get_idle_time(int cpu);
 
 #endif /* _LINUX_SCHED_CPUTIME_H */
diff --git a/include/linux/sched/nohz.h b/include/linux/sched/nohz.h
index 6d67e9a5af6b..8747963d2d4b 100644
--- a/include/linux/sched/nohz.h
+++ b/include/linux/sched/nohz.h
@@ -14,10 +14,12 @@ static inline void nohz_balance_enter_idle(int cpu) { }
 #endif
 
 #ifdef CONFIG_NO_HZ_COMMON
+extern unsigned long tick_nohz_active;
 void calc_load_nohz_start(void);
 void calc_load_nohz_remote(struct rq *rq);
 void calc_load_nohz_stop(void);
 #else
+#define tick_nohz_active (0)
 static inline void calc_load_nohz_start(void) { }
 static inline void calc_load_nohz_remote(struct rq *rq) { }
 static inline void calc_load_nohz_stop(void) { }
diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index c8e029373a09..78bc0d027d22 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -7302,6 +7302,17 @@ static struct kmem_cache *task_group_cache __read_mostly;
 DECLARE_PER_CPU(cpumask_var_t, load_balance_mask);
 DECLARE_PER_CPU(cpumask_var_t, select_idle_mask);
 
+static __init void rq_cputime_init(void)
+{
+	int cpu;
+	struct rq_cputime *rq_cputime;
+
+	for_each_possible_cpu(cpu) {
+		rq_cputime = &per_cpu(rq_cputimes, cpu);
+		raw_spin_lock_init(&rq_cputime->lock);
+	}
+}
+
 void __init sched_init(void)
 {
 	unsigned long ptr = 0;
@@ -7475,6 +7486,9 @@ void __init sched_init(void)
 
 	init_uclamp();
 
+	if (static_branch_unlikely(&use_sched_idle_time))
+		rq_cputime_init();
+
 	scheduler_running = 1;
 }
 
diff --git a/kernel/sched/cputime.c b/kernel/sched/cputime.c
index ca0eef7d3852..5af20da6d249 100644
--- a/kernel/sched/cputime.c
+++ b/kernel/sched/cputime.c
@@ -1078,3 +1078,70 @@ void kcpustat_cpu_fetch(struct kernel_cpustat *dst, int cpu)
 EXPORT_SYMBOL_GPL(kcpustat_cpu_fetch);
 
 #endif /* CONFIG_VIRT_CPU_ACCOUNTING_GEN */
+
+DEFINE_STATIC_KEY_FALSE(use_sched_idle_time);
+DEFINE_PER_CPU_SHARED_ALIGNED(struct rq_cputime, rq_cputimes);
+
+static int __init sched_idle_time_setup(char *str)
+{
+	static_branch_enable(&use_sched_idle_time);
+	return 0;
+}
+early_param("use-sched-idle-time", sched_idle_time_setup);
+
+int sched_idle_time_adjust(int cpu, u64 *utime, u64 *stime)
+{
+	struct rq_cputime *rq_cputime = &per_cpu(rq_cputimes, cpu);
+	struct cputime *prev = &rq_cputime->cpu_prev_time;
+	struct cputime *last = &rq_cputime->cpu_last_time;
+	u64 ut, st, delta, delta_ut, delta_st;
+	struct kernel_cpustat kcs;
+
+	raw_spin_lock(&rq_cputime->lock);
+	kcpustat_cpu_fetch(&kcs, cpu);
+
+	delta = cpu_clock(cpu) - get_idle_time(&kcs, cpu) - (prev->utime + prev->stime);
+	ut = kcpustat_cpu(cpu).cpustat[CPUTIME_USER];
+	st = kcpustat_cpu(cpu).cpustat[CPUTIME_SYSTEM];
+
+	delta_ut = ut - last->utime;
+	delta_st = st - last->stime;
+
+	if (unlikely((s64)delta <= 0))
+		goto out;
+
+	if (delta_st == 0) {
+		prev->utime += delta;
+	} else if (delta_ut == 0) {
+		prev->stime += delta;
+	} else {
+		delta_st = mul_u64_u64_div_u64(delta_st, delta, delta_ut + delta_st);
+
+		if (unlikely(delta_st > delta))
+			delta_st = delta;
+
+		prev->stime += delta_st;
+		prev->utime += delta - delta_st;
+	}
+
+out:
+	last->utime = ut;
+	last->stime = st;
+
+	*utime = prev->utime;
+	*stime = prev->stime;
+
+	raw_spin_unlock(&rq_cputime->lock);
+
+	return 0;
+}
+
+unsigned long long sched_get_idle_time(int cpu)
+{
+	struct rq_cputime *rt = &per_cpu(rq_cputimes, cpu);
+
+	if (is_idle_task(cpu_curr(cpu)))
+		return rt->sum_idle_time + cpu_clock(cpu) - rt->last_entry_idle;
+	else
+		return rt->sum_idle_time;
+}
diff --git a/kernel/sched/idle.c b/kernel/sched/idle.c
index 2593a733c084..17d427e9c775 100644
--- a/kernel/sched/idle.c
+++ b/kernel/sched/idle.c
@@ -424,12 +424,22 @@ static void check_preempt_curr_idle(struct rq *rq, struct task_struct *p, int fl
 
 static void put_prev_task_idle(struct rq *rq, struct task_struct *prev)
 {
+	if (static_branch_unlikely(&use_sched_idle_time) && !tick_nohz_active) {
+		struct rq_cputime *rq_cputime = this_cpu_ptr(&rq_cputimes);
+		unsigned long long now = cpu_clock(cpu_of(rq)), delta = 0;
+
+		delta = now - rq_cputime->last_entry_idle;
+		rq_cputime->sum_idle_time += delta;
+	}
 }
 
 static void set_next_task_idle(struct rq *rq, struct task_struct *next, bool first)
 {
 	update_idle_core(rq);
 	schedstat_inc(rq->sched_goidle);
+
+	if (static_branch_unlikely(&use_sched_idle_time) && !tick_nohz_active)
+		this_cpu_ptr(&rq_cputimes)->last_entry_idle = cpu_clock(cpu_of(rq));
 }
 
 struct task_struct *pick_next_task_idle(struct rq *rq)
diff --git a/kernel/sched/sched.h b/kernel/sched/sched.h
index 35cc7ab9f783..491d530851cc 100644
--- a/kernel/sched/sched.h
+++ b/kernel/sched/sched.h
@@ -1122,6 +1122,21 @@ static inline struct rq *rq_of(struct cfs_rq *cfs_rq)
 }
 #endif
 
+struct cputime {
+	u64 utime;
+	u64 stime;
+};
+
+struct rq_cputime {
+	raw_spinlock_t lock;
+	unsigned long long sum_idle_time;
+	unsigned long long last_entry_idle;
+	struct cputime cpu_prev_time;
+	struct cputime cpu_last_time;
+};
+
+DECLARE_PER_CPU(struct rq_cputime, rq_cputimes);
+
 static inline int cpu_of(struct rq *rq)
 {
 #ifdef CONFIG_SMP
diff --git a/kernel/time/tick-internal.h b/kernel/time/tick-internal.h
index 5294f5b1f955..64d9ace7f14a 100644
--- a/kernel/time/tick-internal.h
+++ b/kernel/time/tick-internal.h
@@ -4,6 +4,7 @@
  */
 #include <linux/hrtimer.h>
 #include <linux/tick.h>
+#include <linux/sched/nohz.h>
 
 #include "timekeeping.h"
 #include "tick-sched.h"
@@ -151,14 +152,12 @@ static inline void tick_nohz_init(void) { }
 #endif
 
 #ifdef CONFIG_NO_HZ_COMMON
-extern unsigned long tick_nohz_active;
 extern void timers_update_nohz(void);
 # ifdef CONFIG_SMP
 extern struct static_key_false timers_migration_enabled;
 # endif
 #else /* CONFIG_NO_HZ_COMMON */
 static inline void timers_update_nohz(void) { }
-#define tick_nohz_active (0)
 #endif
 
 DECLARE_PER_CPU(struct hrtimer_cpu_base, hrtimer_bases);
-- 
2.17.1

